sui_programmability/examples/defi/sources/pool.move

```diff
@@ -0,0 +1,375 @@
+            swap_sui(pool, sui, ctx),
+            TxContext::sender(ctx)
+        )
+    }
+
+    /// Swap `Coin<SUI>` for the `Coin<T>`.
+    /// Returns Coin<T>.
+    public fun swap_sui<T>(
+        pool: &mut Pool<T>, sui: Coin<SUI>, ctx: &mut TxContext
+    ): Coin<T> {
+        assert!(Coin::value(&sui) > 0, EZeroAmount);
+
+        let sui_balance = Coin::into_balance(sui);
+
+        // Calculate the output amount - fee
+        let (sui_reserve, token_reserve, _) = get_amounts(pool);
+        let output_amount = get_input_price(
+            Balance::value(&sui_balance),
+            sui_reserve,
+            token_reserve,
+            pool.fee_percent
+        );
+
+        Balance::join(&mut pool.sui, sui_balance);
+        Coin::withdraw(&mut pool.token, output_amount, ctx)
+    }
+
+    /// Entry point for the `swap_token` method. Sends swapped SUI
+    /// to the sender.
+    public(script) fun swap_token_<T>(
+        pool: &mut Pool<T>, token: Coin<T>, ctx: &mut TxContext
+    ) {
+        Transfer::transfer(
+            swap_token(pool, token, ctx),
+            TxContext::sender(ctx)
+        )
+    }
+
+    /// Swap `Coin<T>` for the `Coin<SUI>`.
+    /// Returns the swapped `Coin<SUI>`.
+    public fun swap_token<T>(
+        pool: &mut Pool<T>, token: Coin<T>, ctx: &mut TxContext
+    ): Coin<SUI> {
+        assert!(Coin::value(&token) > 0, EZeroAmount);
+
+        let tok_balance = Coin::into_balance(token);
+
+        // Calculate the output amount - fee
+        let (sui_reserve, token_reserve, _) = get_amounts(pool);
+        let output_amount = get_input_price(
+            Balance::value(&tok_balance),
+            token_reserve,
+            sui_reserve,
+            pool.fee_percent
+        );
+
+        Balance::join(&mut pool.token, tok_balance);
+        Coin::withdraw(&mut pool.sui, output_amount, ctx)
+    }
+
+    /// Entrypoint for the `add_liquidity` method. Sends `Coin<LSP>` to
+    /// the transaction sender.
+    public(script) fun add_liquidity_<T>(
+        pool: &mut Pool<T>, sui: Coin<SUI>, token: Coin<T>, ctx: &mut TxContext
+    ) {
+        Transfer::transfer(
+            add_liquidity(pool, sui, token, ctx),
+            TxContext::sender(ctx)
+        );
+    }
+
+    /// Add liquidity to the `Pool`. Sender needs to provide both
+    /// `Coin<SUI>` and `Coin<T>`, and in exchange he gets `Coin<LSP>` -
+    /// liquidity provider tokens.
+    public fun add_liquidity<T>(
+        pool: &mut Pool<T>, sui: Coin<SUI>, token: Coin<T>, ctx: &mut TxContext
+    ): Coin<LSP<T>> {
+        assert!(Coin::value(&sui) > 0, EZeroAmount);
+        assert!(Coin::value(&token) > 0, EZeroAmount);
+
+        let sui_balance = Coin::into_balance(sui);
+        let token_balance = Coin::into_balance(token);
+
+        let (sui_amount, _, lsp_supply) = get_amounts(pool);
+
+        let sui_added = Balance::value(&sui_balance);
+        let share_minted = (sui_added * lsp_supply) / sui_amount;
```

Comment:
Do we want a better error for when these various amounts hit 0? (saw in a few other spots)